### PR TITLE
Add support for teams

### PR DESCRIPTION
**Pull request by gerits**
We are developing an application with about 30 developers that all contribute in the same repository. It's not interesting to review everybody's code, so I've added a team functionality. 

It allows to create teams and add existing users to this team.
A team member can either be a spectator or a contributor. The commits of all the contributors are listed in the "to review" tab. Being part of a team as spectator allows you to view and comment, but other members won't be able to see your commits.

These are my initial requirements and I've limited my implementation to this. I was not sure yet on how to display it in the ui and what to do as default behavior for users that are not part of a team (nothing to review?).

Does a functionality like this require a more advanced user system or can it exist separate from the user system as I've implemented it now.

### CODE REVIEW

**Comment by bazhe** [comment]
This is really one interest feature, I will waiting for your merge!

File: run.sh [hunk 1]
```diff
@@ -4,4 +4,4 @@ cd codebrag-ui
 nohup npm install && ./node_modules/.bin/grunt server &
 
 cd ../
-java -Dfile.encoding=UTF8 -Xmx3000M -Xss1M -XX:+CMSClassUnloadingEnabled -XX:MaxPermSize=1024m -jar sbt-launch.jar "~ container:start"
+java -Dfile.encoding=UTF8 -Xmx1590M -Xss1M -XX:+CMSClassUnloadingEnabled -XX:MaxPermSize=1024m -jar sbt-launch.jar "~ container:start"
```
Comment (lukaszlenart) [inline run.sh]: Why did you reduce this?

**Comment by lukaszlenart** [comment]
I'm wondering how to treat teamless users, showing `nothing to review` isn't the best idea. I would rather show everything, at the end a `team` is yet another filter. It would be good to have a switch to disable `team` filter but that can be added latter

File: run.sh (hunk 1 above)
Comment (gerits) [inline run.sh]: I was developing on a 32-bit machine, this has a memory limitation. But this should indeed not be part of the merge. I will revert this.

File: codebrag-rest/src/main/scala/com/softwaremill/codebrag/rest/TeamsServlet.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  get("/:teamId/members") {
+    haltIfNotAuthenticated()
+    if (!config.demo) {
+      var team = teamFinder.findTeam(new ObjectId(params("teamId")));
+      teamFinder.findAllAsManagedTeamMembers(team)
+    } else {
+      ManagedTeamMembersListView(List.empty)
+    }
```
Comment (lukaszlenart) [inline codebrag-rest/src/main/scala/com/softwaremill/codebrag/rest/TeamsServlet.scala]: Could you revert this `if` clause?
```scala
if (config.demo) {
  ManagedTeamMembersListView(List.empty)
} else {
  var team = teamFinder.findTeam(new ObjectId(params("teamId")));
  teamFinder.findAllAsManagedTeamMembers(team)
}
```

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/finders/commits/UserLoader.scala [hunk 3]
```diff
@@ -9,5 +9,5 @@ protected[finders] trait UserLoader extends Logging {
   protected def userDao: UserDAO
 
   protected def loadUser(userId: ObjectId) = userDao.findById(userId).getOrElse(throw new IllegalArgumentException("Invalid userId provided"))
-
+
```
Comment (lukaszlenart) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/finders/commits/UserLoader.scala]: Please don't :)

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/finders/commits/all/AllCommitsFinder.scala [hunk 4]
```diff
@@ -29,6 +29,6 @@ class AllCommitsFinder(
       case None => Left("Commit not found")
     }
   }
-
+
```
Comment (lukaszlenart) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/finders/commits/all/AllCommitsFinder.scala]: Same here, please don't :)

File: codebrag-service/src/main/scala/com/softwaremill/codebrag/usecases/team/ModifyTeamMemberUseCase.scala [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import com.softwaremill.codebrag.dao.user.TeamDAO
+import com.softwaremill.codebrag.domain.Team
+import com.softwaremill.codebrag.domain.TeamMember
+
+class ModifyTeamMemberUseCase(val teamDao: TeamDAO) extends Logging {
+
+  def execute(teamId: ObjectId, userId: ObjectId, contributor: Option[Boolean]): Either[Errors, Team] = {
+    validateTeam(teamId).whenOk {
+      var team = teamDao.findById(teamId).get
+      if (contributor != None) {
```
Comment (lukaszlenart) [inline codebrag-service/src/main/scala/com/softwaremill/codebrag/usecases/team/ModifyTeamMemberUseCase.scala]: `contributor.isDefined` ?

**Comment by gerits** [comment]
The filtering of the team members seemed logical at first for an all team setup. But it indeed requires some visualisation:
- By default show everything. 
- When in a team there should be somewhere an indication of which team (profile page?). 
- Extra filter needed?: to review team members, to review all, all commits
- Possibility to request to join a team or fully managed by admin.

**Comment by lukaszlenart** [comment]
Let's start with the simplest option - user is not assigned to any team, show everything

**Comment by lukaszlenart** [comment]
Is it ready for final shot?

**Comment by gerits** [comment]
We have been using it for the past weeks now, and did not see any problems. Although addition features could be added, this is a stable first version of the 'teams' support.

**Comment by lukaszlenart** [comment]
Gut! Waiting for final confirmation from teammates and

**Pull request merged**